### PR TITLE
Feat/buses crud

### PR DESCRIPTION
**Pull request by IrakozeLoraine**
## What does this PR do
- It implements the bus feature
- It adds the feature tests

## Description
- Operator is able to perform the CRUD operations on buses
- Buses can be seeded at once for development purposes by running `npm run db:seeds`

## How to manually test this 
- Check in the branch `feat/buses-crud`
- Run the server
- Open the Swagger API documentation to test the buses

## Relevant Trello story 
[11-crud-operations-for-buses][(https://trello.com/c/a6wcZ3PE/9-create-roles-and-set-permissions)](https://trello.com/c/maUQgDhP/11-crud-operations-for-buses)

## Screenshots
image

image

### CODE REVIEW

File: src/api/controllers/bus.controllers.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    try {
+        const bus = await findBusById(req.params.id)
+        
+        if (!bus) {
+            return res.status(StatusCodes.NOT_FOUND).json({
+                success: false,
+                message: locales('bus_not_found'),
+            });
+        }
+
```
Comment (TuyizeeAnastase) [inline src/api/controllers/bus.controllers.js]: add middleware which check this, check if bus exist by id

File: src/api/controllers/bus.controllers.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        })
+    } catch (err) {
+       next(err)
+    }
+}
+
+export const deleteBusById = async (req, res, next) => {
+    try {
+        const bus = await findBusById(req.params.id)
+        if (!bus) {
```
Comment (TuyizeeAnastase) [inline src/api/controllers/bus.controllers.js]: use above option using middleware to check if bus exist by id

File: src/api/controllers/bus.controllers.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    } catch (err) {
+        next(err)
+    }
+}
+
+export const updateBus = async (req, res, next)=> {
+    try {
+        await editBus(req.body, req.params.id)
+        
+        res.status(StatusCodes.OK).json({
```
Comment (TuyizeeAnastase) [inline src/api/controllers/bus.controllers.js]: the same like above add middleware to check if bus exist by id

File: src/api/services/bus.service.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    try {
+      const data = await Bus.findAll();
+      return data;
+    } catch (e) {
+      throw new CustomError(
+        e?.message || "Error fetching buses",
+        StatusCodes.INTERNAL_SERVER_ERROR
+      );
+    }
+};
```
Comment (TuyizeeAnastase) [inline src/api/services/bus.service.js]: add pagination to get All buses to refer to this [pagination](https://www.bezkoder.com/node-js-sequelize-pagination-mysql/)

File: src/api/validations/bus.validate.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import Joi from "joi";
+import { StatusCodes } from "http-status-codes";
+
+export const validateCreateBus = (req, res, next) => {
+  const schema = Joi.object({
+    plate_number: Joi.string().required(),
+    driver_id: Joi.string().required(),
+    agency_id: Joi.string().required(),
```
Comment (TuyizeeAnastase) [inline src/api/validations/bus.validate.js]: driver_id and agency_id are not string

File: src/db/migrations/20230323144307-create-driver.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        allowNull: false,
+        type: Sequelize.DATE,
+      },
+      updatedAt: {
+        allowNull: false,
+        type: Sequelize.DATE,
+      },
+    });
+  },
+  async down(queryInterface, Sequelize) {
```
Comment (TuyizeeAnastase) [inline src/db/migrations/20230323144307-create-driver.js]: we don't have a table called Driver it's User remove it,

File: src/db/seeders/20230323144350-Agency.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
+'use strict';
```
Comment (TuyizeeAnastase) [inline src/db/seeders/20230323144350-Agency.js]: add agencies seeding

File: src/api/routes/buses.routes.js [hunk 8]
```diff
@@ -111,25 +112,14 @@ router.get("/:id", (req, res) => {
...
-    };
-
-    req.app.db.get("buses").push(bus).write();
-
-    res.send(bus);
-  } catch (error) {
-    return res.status(500).send(error);
-  }
-});
+router.post("/", createBus)
```
Comment (TuyizeeAnastase) [inline src/api/routes/buses.routes.js]: before adding the bus add this middleware
- check if the bus is not registered by plate_number
- check if user Exists and it has driver as rols
- check if agency exist

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

File: src/api/routes/buses.routes.js (hunk 8 above)
Comment (IrakozeLoraine) [reply src/api/routes/buses.routes.js]: I can't create the middleware for agency until the controllers are ready.

File: src/api/controllers/bus.controllers.js
Comment (TuyizeeAnastase) [inline src/api/controllers/bus.controllers.js:1]: To update and delete the bus add middleware to check if the bus exists by id no need after adding that middleware
`if (!bus) {
            return res.status(StatusCodes.NOT_FOUND).json({
                success: false,
                message: locales('bus_not_found'),
            });
        }`

File: src/api/routes/buses.routes.js [hunk 9]
```diff
@@ -118,15 +152,16 @@ router.get("/:id", getBusById);
  *         description: Some server error
  */
 
-// eslint-disable-next-line consistent-return
-router.post("/", createBus)
+router.post("/", checkUserLoggedIn, restrictTo(ERoles.OPERATOR), busExistsByPlateNumber, createBus)
```
Comment (TuyizeeAnastase) [inline src/api/routes/buses.routes.js]: add middleware to check if agency exist by agency_id

**Review by TuyizeeAnastase — CHANGES_REQUESTED** [review]

**Review by thepatrickniyo — COMMENTED** [review]
There is an error in migration associations. Pull the latest version on develop then if you have time we look into it.

**Comment by SauveJeanLuc** [comment]
resolve conflicts

**Review by thepatrickniyo — APPROVED** [review]

**Review by ernesteNtezirizaza — APPROVED** [review]

**Review by TuyizeeAnastase — APPROVED** [review]

**Pull request merged**